Convert TroopJS example

mikaelkaron opened the pull request:
Hi.

We've started work on converting and think we have something we can share. Note that this is still preliminary code, but we think it's in a state where we are ready for feedback.

These are know issues:

 - bower used for packaging vs. npm. Hope this is not a problem.

These are outstanding tasks:

- [X] Update "completed count" (pending tastejs/todomvc#1215) [_code is there to update text, hopefylly this will work in the future_]
- [x] Add source comments
- [x] Update readme
- [x] Remove unused vendor files

mikaelkaron commented:
At the same time, can TroopJS be move out of "Labs" (we've been part of the project since the initial conception more or less), and if not, what is the motivation to keep it there?

arthurvr commented:
Thanks for your PR @mikaelkaron. I'll review it this evening.

arthurvr commented on examples/troopjs_require/bower.json:
```diff
@@ -1,11 +1,18 @@
-		"troopjs-bundle" : "https://github.com/troopjs/troopjs-bundle/archive/build/2.x.zip"
+	"name": "todos-troopjs",
+	"private": true,
+	"main": "index.html",
```
Moot. We only need `private` and the deps.

arthurvr commented on examples/troopjs_require/bower.json:
```diff
@@ -1,11 +1,18 @@
-		"requirejs" : "~2.1",
-		"jquery" : "~2.0",
-		"troopjs-bundle" : "https://github.com/troopjs/troopjs-bundle/archive/build/2.x.zip"
+	"name": "todos-troopjs",
```
Moot.

arthurvr commented on examples/troopjs_require/bower_components/jquery/.bower.json:
```diff
@@ -0,0 +1,37 @@
+  "_source": "git://github.com/jquery/jquery.git",
+  "_target": "^2",
+  "_originalSource": "jquery"
+}
```
You shouldn't commit all vendor files. Only the ones needed to run the app. All the rest should be git-ignored. Example `.gitignore`: https://github.com/tastejs/todomvc/blob/master/examples/spine/.gitignore.

arthurvr commented on examples/troopjs_require/bower.json:
```diff
@@ -1,11 +1,18 @@
+		"when": "^3.4"
+	},
+	"devDependencies": {
+		"requirejs": "^2"
```
The site depends on this so this should be in `dependencies`.

arthurvr commented on examples/troopjs_require/bower.json:
```diff
@@ -1,11 +1,18 @@
+		"troopjs-route-hash": "^1",
+		"jquery": "^2",
+		"poly": "^0.6",
+		"when": "^3.4"
```
These are tabs. [`package.json` should use 2 spaces.](https://github.com/tastejs/todomvc/commit/64484b9e18eabed01b8314af899f50697d167946)

arthurvr commented:
> At the same time, can TroopJS be move out of "Labs" (we've been part of the project since the initial conception more or less), and if not, what is the motivation to keep it there?

That will hopefully get discussed quick: https://github.com/tastejs/todomvc/issues/845.

addyosmani commented:
My view is that if it meets our specs and passes tests, let's move it out of labs.

passy commented:
+1

samccone commented:
(and it has an active maintainer)  :+1:

arthurvr commented:
I'm +1 as well. Doing it right now :D

arthurvr commented:
Done. Thanks @mikaelkaron!

arthurvr commented:
> (pending tastejs/todomvc#1215)

That shouldn't really hold us up. We've been updating 45 other apps as well :D Visually everything works fine so let's not wait for the tests/css to get fixed.

arthurvr commented:
>  Update tagline used on homepage, Update readme,  Add source comments

Awesome you're doing it but it's better to have these as different PRs. Keeping stuff focussed, you know.

mikaelkaron commented:
Ok. All the changes related to source changes is now completed (I included the change to `readme.md` as it had a new format in the template).

I'd still like to update (in separate PR, in the near future)

* Directory name from `troopjs_require` -> `troopjs`
* Tagline used on frontpage

But non of those should block this PR

arthurvr commented:
Just wondering, Is there something framework-specific holding us up to switch to npm?

samccone commented:
:+1: to move to npm, since that it where we are pushing all the other examples

mikaelkaron commented:
At the moment neither TroopJS or it's dependencies is available on npm.

Our team has a lot more experience with bower so that's what we're supporting out of the box. We're actually looking at the effort needed to support npm as a first class citizen, primarily reading up on [what the npm blog has to says about it](http://blog.npmjs.org/post/101775448305/npm-and-front-end-packaging).

Could we perhaps proceed with this PR since at the moment bower is "best practice" as far as the framework is concerned, and _if_ we decide to support npm we could amend with a new PR?

sindresorhus commented:
npm supports git deps (which is what bower uses), so it supports anything bower does.

samccone commented:
*caveat*
npm will __not__ work if there is no package.json, bower will

you can use napa to get around this like we do here
https://github.com/tastejs/todomvc/blob/master/examples/canjs/package.json#L11

mikaelkaron commented:
So I've read [the npm blog](http://blog.npmjs.org/post/101775448305/npm-and-front-end-packaging), skimmed the [i-wanna-be-the-very-best label for npm issues](https://github.com/npm/npm/labels/i-wanna-be-the-very-best), studied [bower ticket #1520](https://github.com/bower/bower/issues/1520) and a [few](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) [other](https://docs.npmjs.com/cli/dedupe) [resources](http://blog.npmjs.org/post/111475741445/publishing-your-jquery-plugin-to-npm-the-quick) and frankly I don't see how the recursive nature of `node_modules` can ever be compatible with the flat structure of `bower_components`. There are workarounds using [browserify](http://browserify.org/), but TroopJS truly depends on the `A` in `AMD` (asynchronous) so unfortunately that's not going to work for us.

It _could_ be possible to create a `package.json` file that would pull each dependency for this specific app and put them in a predictable flat tree but it feels like we're just using `npm` for the sake of using `npm` rather than "best tool for the job" (IMHO).

However, if someone can point me in the direction of how to support a flat dependency tree using `npm` with deep (duplicate) dependencies I'd be happy to try it out and if it's not _to_ much work make all the TroopJS stuff available on `npm`. I've taken a brief look at the [jquery npm package](https://www.npmjs.com/package/jquery) and it's dependents and I don't see anyone who solves this in an elegant way, but maybe I'm just looking in the wrong place?

mikaelkaron commented:
It looks like [I'm not the only one](http://stackoverflow.com/a/18652918/232488) concerned with using `npm` for the front-end and the problem of `node_modules` and deep dependencies... Granted it's been a while since that post but I don't see how things have changed for the better since then.

mikaelkaron commented:
@samccone If we were to use `napa` for this we'd have to create a `package.json` that looks something like this:

```json
{
  "dependencies": {
    "jquery": "^2",
    "when": "^3.4",
    "poly": "git://github.com/cujojs/poly#0.6.0",
    "requirejs": "^2.1",
    "todomvc-app-css": "^2",
    "todomvc-common": "^1"
  },
  "devDependencies": {
    "napa": "^1.2.0"
  },
  "scripts": {
    "install": "napa"
  },
  "napa": {
    "mu-emitter": "mu-lib/mu-emitter#0.0.8",
    "mu-getargs": "mu-lib/mu-getargs#0.0.3",
    "mu-merge": "mu-lib/mu-merge#0.0.2",
    "mu-selector-set": "mu-lib/mu-selector-set#1.1.0",
    "mu-unique": "mu-lib/mu-unique#0.0.5",
    "troopjs": "troopjs/troopjs#3.0.0-rc.5",
    "troopjs-route-hash": "troopjs/troopjs-route-hash#1.0.0-rc.3",
    "troopjs-widget": "troopjs/troopjs-widget#1.0.0-rc.4.2"
  }
}
```

However it feels like all we're doing is to use `npm` as a fancy downloader and we may as well have created a `package.json` like this:

```json
{
  "devDependencies": {
    "bower": "^1.3"
  },
  "scripts": {
    "install": "bower install"
  }
}
```

mikaelkaron commented:
According to [npmweekly 5](http://blog.npmjs.org/post/110924823920/npm-weekly-5) `node_modules` will be flatter in `node@3`:

> With npm@3, your `node_modules` directory will be a lot flatter. All of your dependencies and most of your subdependencies (and (sub)+dependencies) will be sitting next to each other at the top level. Only when there are conflicts will modules be installed at deeper levels.

This is a good start but to make it "as good as bower" for dependency conflict resolution `npm` would have to _ask_ what version of a dependency to install during conflict rather than simply nesting it. Glancing at the [npm roadmap](https://github.com/npm/npm/wiki/Roadmap) it's apparent that the `npm` developers are well aware of the issues and there's even a [separate wiki page](https://github.com/npm/npm/wiki/npm%E2%87%94Angular-brainstorming-session) for npm⇔Angular integration.

However, none of these solutions are available today, and perhaps this is not the place to raise a discussion like this, but from what I can see `npm` simply is not ready for broad usage in some specific (albeit common) scenarios (like `AMD` + browser), so why is it being promoted as such?

mikaelkaron commented:
Looks like I effectively killed the conversation there.

If you guys are OK with it I'll just add a `package.json` structured like above in a new commit and hopefully we can move this PR along. Sorry for derailing the PR into a `npm` vs. `bower` conversation.

mikaelkaron commented:
This has been quiet for a while now, is there something outstanding to merge?

mikaelkaron commented:
@arthurvr @samccone ping! Just wanted to know if there's anything else I can do to get this PR merged?

arthurvr commented on examples/troopjs_require/index.html:
```diff
@@ -1,77 +1,56 @@
-	</body>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
```
We use tab indentation.

arthurvr commented:
@tastejs/todomvc This dependency system seems like overcomplicating stuff. What do we think?

arthurvr commented on examples/troopjs_require/readme.md:
```diff
@@ -1,16 +1,21 @@
-# TroopJS TodoMVC Example
+# TroopJS • [TodoMVC](http://todomvc.com)
```
Could you please keep unrelated changes (like the ones in this file) in separated commits? That helps us reviewing and keeps the commit log clean.

mikaelkaron replied:
All changes for `readme.md` are already in a separate commit. Did you mean separate PR?

arthurvr commented on examples/troopjs_require/package.json:
```diff
@@ -0,0 +1,29 @@
+    "when": "^3.4",
+    "poly": "git://github.com/cujojs/poly#0.6.0",
+    "requirejs": "^2.1",
+    "todomvc-app-css": "^2",
```
@tastejs/todomvc This app already updated to classes, which is good, but we will need to wait to land the PR before the tests are updated right?

samccone replied:
unless we change all the tests... I would rather wait to land this change only increasing the spread of implementations :grimacing:

arthurvr replied:
Yeah I thought so.

mikaelkaron commented:
So I:
 - Rebased against tastejs:master
 - Made sure related changes are separated in commits
    - readme [5d14504]
    - source [0fbb24a]
    - comments [5b4f0d3]
    - packaging [c909e6c] [1387bc1]
 - Made sure to use latest `index.html` template from [here](https://github.com/tastejs/todomvc-app-template/blob/master/index.html)
 - Validated JS code with [jscs](jscs.info) with config from [here](https://github.com/tastejs/todomvc/blob/master/.jscsrc)

The main discussion of this PR looks to me to have been related to `npm` vs. `bower` (sorry for that) so I've tried to contain that change in a separate commit [1387bc1] which can be easily reverted if needed.

Is there anything more I can do to move this along?

arthurvr commented on examples/troopjs_require/index.html:
```diff
@@ -1,77 +1,63 @@
 		<title>TroopJS • TodoMVC</title>
-		<link rel="stylesheet" href="bower_components/todomvc-common/base.css">
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+		<!-- CSS overrides - remove if you don't need it -->
```
remove this comment

arthurvr commented:
This needs a rebase.

arthurvr commented on examples/troopjs_require/js/clear.js:
```diff
@@ -0,0 +1,40 @@
+
+			// Update text and toggle visibility (visible if `count > 0`, hidden otherwise)
+			this.$element
+				.text('Clear completed (' + count + ')')
```
This count should be removed: https://github.com/tastejs/todomvc/issues/1234.

arthurvr commented:
Fix those things and this should be good to go.

samccone commented:
@mikaelkaron sorry about the delay here, would you mind rebasing off of master and pushing again, we recently made the tests work with both version of the spec :+1: :volcano: 

It also looks like you are running into a few style related issues, please refer to the built errors.

Thanks again :sparkling_heart:

mikaelkaron commented:
I rebased and fixed lint errors, is there anything else I can do.

I'd also like to restate the question if using `npm` for dependency management is a requirement? I personally prefer `bower` and I'd like people who are thinking of developing TroopJS applications to think about using the same. I do understand the argument that for the TodoMVC project you'd like everyone to be as close to a baseline as possible and that baseline happens to be `npm`.

If that is indeed the case it would be nice to have an official FAQ ticket or similar explaining the stance on the issue. At the moment it feels a little bit arbitrary.

mikaelkaron commented:
Looks like there are still tests failing, strange. Did something change?

sindresorhus commented:
I reran the Travis build and now got:

```
npm ERR! Linux 3.13.0-40-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v0.12.7/bin/node" "/home/travis/.nvm/versions/node/v0.12.7/bin/npm" "run" "lint" "--" "-c" "../.jscsrc" "../examples/gwt" "../examples/troopjs_require"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! code ELIFECYCLE
npm ERR! @ lint: `jscs "-c" "../.jscsrc" "../examples/gwt" "../examples/troopjs_require"`
npm ERR! Exit status 134
npm ERR! 
npm ERR! Failed at the @ lint script 'jscs "-c" "../.jscsrc" "../examples/gwt" "../examples/troopjs_require"'.
npm ERR! This is most likely a problem with the  package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     jscs "-c" "../.jscsrc" "../examples/gwt" "../examples/troopjs_require"
npm ERR! You can get their info via:
npm ERR!     npm owner ls 
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/tastejs/todomvc/tooling/npm-debug.log
```

@samccone Thoughts?

sindresorhus commented:
> I'd also like to restate the question if using npm for dependency management is a requirement?

Yes

> If that is indeed the case it would be nice to have an official FAQ ticket or similar explaining the stance on the issue. At the moment it feels a little bit arbitrary.

Yup, we should.

sindresorhus commented on examples/troopjs_require/.gitignore:
```diff
@@ -0,0 +1,51 @@
+!node_modules/mu-selector-set/src/matchesSelector.js
+!node_modules/mu-selector-set/src/classifier.js
+!node_modules/mu-selector-set/src/Subsets.js
+!node_modules/mu-selector-set/src/splitter.js
```
You can probably use a glob pattern here `!node_modules/mu-selector-set/src/*`, unless you use any files in the `src` folder.

sindresorhus commented on examples/troopjs_require/index.html:
```diff
@@ -1,77 +1,62 @@
 <!doctype html>
-<html lang="en" data-framework="troopjs">
+<html lang="en">
```
Why are you removing this?

sindresorhus commented on examples/troopjs_require/index.html:
```diff
@@ -1,77 +1,62 @@
+				<h1>todos</h1>
+				<input class="new-todo" placeholder="What needs to be done?" autofocus data-weave="todos-troopjs/create">
+			</header>
+			<!-- This section should be hidden by default and shown when there are todos -->
```
As per spec, default comments should be removed. Applies to the other ones too.

sindresorhus commented on examples/troopjs_require/js/clear.js:
```diff
@@ -0,0 +1,37 @@
+		     */
+		'hub/todos/change(true)': function (tasks) {
+			// Filter and count tasks that are completed
+			var count = tasks
```
semicolon

sindresorhus commented on examples/troopjs_require/js/clear.js:
```diff
@@ -0,0 +1,37 @@
+        .filter(function (task) {
+	return task.completed;
+        })
+        .length;
```
Weird indentation

sindresorhus commented on examples/troopjs_require/js/count.js:
```diff
@@ -0,0 +1,29 @@
+				.length;
+
+			// Update HTML with `count` taking pluralization into account
+			this.$element.html('<strong>' + count + '</strong> ' + (count === 1 ? 'item' : 'items') + ' left');
```
You can remove the `<strong>`. It's no longer used.

sindresorhus commented on examples/troopjs_require/js/filter.js:
```diff
@@ -0,0 +1,41 @@
+			return hub
+				// Emit `todos/filter` with the new `route` on `hub`, yield `void 0` to not mutate arguments for next handler
+				.emit('todos/filter', route)
+				.yield(void 0);
```
`void 0` => `undefined`

sindresorhus replied:
Actually `.yield();` will do the same.

sindresorhus commented on examples/troopjs_require/js/list.js:
```diff
@@ -0,0 +1,288 @@
+
+			// Call `update`, yield `void 0` to not mutate arguments for next handler
+			return update.call(this)
+				.yield(void 0);
```
https://github.com/tastejs/todomvc/pull/1218/files#r37666252

samccone commented:
friendly ping @mikaelkaron

mikaelkaron commented:
@samccone friendly pong. I'm a bit swamped atm, but this is no.1 non work priority so I hope to get to it in a few days.

samccone commented:
Awesome thanks. No rush just checking in 👏👌😎

samccone commented:
My friendly 2 week :white_check_mark: in :)

samccone commented:
Hey @mikaelkaron I am going to close this PR, when you are ready to pick it back up just let us know and we can reopen the PR.

mikaelkaron commented:
sounds fair. I'll try to push an update and reopen

samccone commented:
sounds good to me, thanks for the update @mikaelkaron :)

arthurvr commented:
Friendly ping.

> At the same time, can TroopJS be move out of "Labs" (we've been part of the project since the initial conception more or less), and if not, what is the motivation to keep it there?

The fact that the app would be up to date and we had an active maintainer was one of the reasons we moved it out of the labs so if those things both fall away I think we should reconsider that decision.

mikaelkaron commented:
I guess I better get to fixing this... I'm sort of on an internet hiatus wich is limiting how much I code, but I should be able to dust this off and finish it.

mikaelkaron commented:
I've updated code, mind reopening?

samccone commented:
Hey @mikaelkaron go ahead and open a new PR and just ink back here.  Thanks!
